node/test/chan_drain.js

```diff
@@ -0,0 +1,374 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var collectParallel = require('collect-parallel/array');
+
+var allocCluster = require('./lib/alloc-cluster.js');
+
+allocCluster.test('immediate drain', {
+    numPeers: 4,
+    skipEmptyCheck: true
+}, function t(cluster, assert) {
+    var server = cluster.channels[0];
+    setupTestClients(cluster, ['a'], runTest);
+    setupServiceServer(server, 'a', 5);
+
+    function runTest(err) {
+        if (err) {
+            finish(err);
+            return;
+        }
+
+        assert.timeoutAfter(20);
+        server.drain('testdown', drained);
+    }
+
+    function drained() {
+        assert.pass('immediate drain happened');
+        server.close(closed);
+    }
+
+    function closed(err) {
+        if (err) {
+            finish(err);
+            return;
+        }
+
+        assert.pass('server closed');
+        setTimeout(finish, 1);
+    }
+
+    function finish(err) {
+        checkFinish(assert, err, cluster, 1);
+    }
+});
+
+allocCluster.test('drain server with a few incoming', {
+    numPeers: 4,
+    skipEmptyCheck: true
+}, function t(cluster, assert) {
+    var server = cluster.channels[0];
+    var clients = null;
+    var finishCount = 0;
+    var reqN = 0;
+    setupTestClients(cluster, ['a'], runTest);
+    setupServiceServer(server, 'a', 5);
+
+    function runTest(err, gotClients) {
+        if (err) {
+            finish(err);
+            return;
+        }
+        clients = gotClients;
+
+        finishCount = 2;
+        assert.timeoutAfter(50);
+        collectParallel(clients.a, sendOne, sendsDone);
+
+        setTimeout(testdown, 1);
```

Comment:
The guaranteed ordering feature of node core timers (which makes it slow) is what makes this 1ms vs 5ms code work.